### PR TITLE
Support Add/Update/Delete events for Pod/Namespace/Network policy.

### PR DESCRIPTION
**Pull request by saiyan86**
**What this PR does / why we need it**:
This PR handles Add/Update/Delete events for Pod/Namespace/Network policy.

**Special notes for your reviewer**:
The functionality has been tested, but not under full coverage.
The test files will be added soon.

### CODE REVIEW

File: cni/npm/ipsm/ipsm.go [hunk 1]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
-func (ipsMgr *IpsetManager) Create(setName string) error {
-	_, exists := ipsMgr.entryMap[setName]
+func isNsSet(setName string) bool {
+	return !strings.Contains(setName, "-") && !strings.Contains(setName, ":")
+}
+
+// CreateList creates an ipset list. npm maintains one setlist per namespace label.
+func (ipsMgr *IpsetManager) CreateList(listName string) error {
+	// Ignore system pods.
+	if listName == "kube-system" {
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: make a "kube-system" constant

File: cni/npm/ipsm/ipsm.go [hunk 2]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+	}
+
+	hashedName := azureNpmPrefix + util.Hash(listName)
+	_, exists := ipsMgr.listMap[listName]
+	if exists {
+		return nil
+	}
+
+	ipsMgr.entryMap[listName] = &ipsEntry{
+		operationFlag: "-N",
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: use one of the "-N" const variables

File: cni/npm/ipsm/ipsm.go [hunk 3]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+
+	ipsMgr.entryMap[listName].operationFlag = ipsetAppendFlag
+	ipsMgr.entryMap[listName].spec = azureNpmPrefix + util.Hash(setName)
+
+	if err := ipsMgr.Run(ipsMgr.entryMap[listName]); err != nil {
+		fmt.Printf("Error creating ipset rules.\n")
+		fmt.Printf("rule: %+v\n", ipsMgr.entryMap[listName])
+		return err
+	}
+	ipsMgr.listMap[listName] = append(ipsMgr.listMap[listName], setName)
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: add space after if statement

File: cni/npm/ipsm/ipsm.go [hunk 4]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+func (ipsMgr *IpsetManager) CreateSet(setName string) error {
+	// Use hashed string for set name to avoid string length limit of ipset.
+	hashedName := azureNpmPrefix + util.Hash(setName)
+	_, exists := ipsMgr.setMap[setName]
+	if exists {
+		return nil
+	}
+
+	ipsMgr.entryMap[setName] = &ipsEntry{
+		operationFlag: "-N",
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: use one of the "-N" const variables

File: cni/npm/ipsm/ipsm.go [hunk 5]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+		set:           hashedName,
+		spec:          "nethash",
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: use a const "nethash"

File: cni/npm/ipsm/ipsm.go [hunk 6]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+	}
+
+	ipsMgr.setMap[setName] = []string{}
+
+	return nil
+}
+
+// AddToSet inserts an ip to an entry in setMap, and creates/updates the corresponding ipset.
+func (ipsMgr *IpsetManager) AddToSet(setName string, ip string) error {
+	if ipsMgr.Exists(setName, ip, "nethash") {
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: use a const "nethash"

File: cni/npm/ipsm/ipsm.go [hunk 7]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
-	return isEmpty, nil
+	return nil
 }
 
 // DeleteSet removes a set from ipset.
 func (ipsMgr *IpsetManager) DeleteSet(setName string) error {
-	entry := &ipsEntry{
-		operationFlag: "-D",
-		set:           setName,
+	_, exists := ipsMgr.setMap[setName]
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go]: change to...

if _, exists := ipsMgr.setMap[setName]; !exists {
&nbsp;&nbsp;&nbsp;&nbsp;return fmt.Errorf("ipset with name %s not found", setName)
}

File: cni/npm/ipsm/ipsm.go [hunk 8]
```diff
@@ -35,115 +66,255 @@ func (ipsMgr *IpsetManager) Exists(key string, val string) bool {
...
+	cmdName := ipset
+	cmdArgs := []string{entry.operationFlag}
+	if len(entry.set) > 0 {
+		cmdArgs = append(cmdArgs, entry.set)
+	}
+	if len(entry.spec) > 0 {
+		cmdArgs = append(cmdArgs, entry.spec)
+	}
+
 	var (
```
Comment (jaer-tsun) [inline cni/npm/ipsm/ipsm.go:291]: declare and initialize in if, then print cmdOut in else-statement

File: cni/npm/iptm/iptm.go [hunk 9]
```diff
@@ -3,90 +3,150 @@ package iptm
...
+	entry.Specs = []string{IptablesJumpFlag, IptablesAzureChain}
+	if err := iptMgr.Run(entry); err != nil {
+		fmt.Printf("Error adding AZURE-NPM chain to FORWARD\n")
+		return err
+	}
+
+	// Add default rule to AZURE-NPM chain.
+	entry.Chain = IptablesAzureChain
+	entry.Specs = []string{
+		"-m",
```
Comment (jaer-tsun) [inline cni/npm/iptm/iptm.go]: assign as constants lines 100-102

File: cni/npm/iptm/iptm.go [hunk 10]
```diff
@@ -3,90 +3,150 @@ package iptm
...
+// Add creates an entry in entryMap, and add corresponding rule in iptables.
+func (iptMgr *IptablesManager) Add(entry *IptEntry) error {
+	// Create iptables rules for every entry in the entryMap.
+	iptMgr.OperationFlag = IptablesAppendFlag
+	fmt.Printf("%+v\n", entry)
+	if err := iptMgr.Run(entry); err != nil {
+		fmt.Printf("Error creating iptables rules.\n")
+		return err
 	}
+	return nil
```
Comment (jaer-tsun) [inline cni/npm/iptm/iptm.go]: leave space after if

File: cni/npm/iptm/iptm.go [hunk 11]
```diff
@@ -3,90 +3,150 @@ package iptm
...
 }
 
 // Run execute an iptables command to update iptables.
-func (iptMgr *IptablesManager) Run(entry *iptEntry) error {
-	cmdName := "iptables"
-	cmdArgs := append([]string{iptMgr.operationFlag, entry.chain}, entry.specs...)
+func (iptMgr *IptablesManager) Run(entry *IptEntry) error {
+	cmdName := iptables
+	cmdArgs := append([]string{iptMgr.OperationFlag, entry.Chain}, entry.Specs...)
 	var (
```
Comment (jaer-tsun) [inline cni/npm/iptm/iptm.go]: declare inside if-statement and print in an else-statement

File: cni/npm/namespace.go [hunk 12]
```diff
@@ -12,64 +12,133 @@ import (
...
+		ipsMgr: ipsm.NewIpsetManager(),
+		iptMgr: iptm.NewIptablesManager(),
 	}
 
 	return ns, nil
 }
 
-// AddNamespace handles add name space.
+func isSystemNs(nsObj *corev1.Namespace) bool {
+	return nsObj.ObjectMeta.Name == "kube-system"
```
Comment (jaer-tsun) [inline cni/npm/namespace.go]: assign const "kube-system"

File: cni/npm/ipsm/ipsm.go (hunk 1 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/ipsm/ipsm.go (hunk 2 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/ipsm/ipsm.go (hunk 3 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/ipsm/ipsm.go (hunk 4 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/ipsm/ipsm.go (hunk 5 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/ipsm/ipsm.go (hunk 6 above)
Comment (saiyan86) [reply cni/npm/ipsm/ipsm.go]: done.

File: cni/npm/namespace.go (hunk 12 above)
Comment (saiyan86) [reply cni/npm/namespace.go]: done.

**Pull request merged**